lib/amqp_client.js

```diff
@@ -219,14 +224,36 @@ AMQPClient.prototype.connect = function(url) {
   });
 };
 
-AMQPClient.prototype._resolveDeferredSenderAttaches = function(linkName, err, link) {
+/**
+ * Return a name and hash for a link associated with an address and options
+ *
+ * @param type      type of link being accessed/created
+ * @param address   the address of the link
+ * @param options   the link creation options
+ * @returns {Object}
+ */
+AMQPClient.prototype._linkInfo = function(type, address, options) {
+  var lookup = (type === 'receiver') ?
+    this._linkNameForHash.receivers : this._linkNameForHash.senders;
+
+  var hash = u.linkHash(address, options);
+  if (lookup.hasOwnProperty(hash)) {
+    return { name: lookup[hash], hash: hash };
+  }
+
+  var name = u.linkName(address, options);
+  lookup[hash] = name;
+  return { name: name, hash: hash };
+};
+
+AMQPClient.prototype._resolveDeferredSenderAttaches = function(linkHash, err, link) {
   // resolve deferred attachments
-  while (this._senderAttaching[linkName].length) {
-    var deferredAttach = this._senderAttaching[linkName].shift();
+  while (this._senderAttaching[linkHash].length) {
```

Comment:
... and then for any map indexed by hash add ByHash to the name

Otherwise, this is just a bug-farm.